src/api/plugins/tdbp/warren_tdbp/indicators.py

```diff
@@ -0,0 +1,550 @@
+
+        # Instantiate indicator
+        sliding_window = SlidingWindowIndicator(
+            window=Window(since=self.until, until=self.until)
+        )
+        active_actions = pd.DataFrame(columns=["iri", "name", "module_type"])
+
+        while since >= min_datetime:
+            # Filter on statements emitted within the sliding window
+            window_statements = statements[since <= statements["date"]]
+
+            # Count unique active students in the sliding window
+            cohort = window_statements["actor.account.name"].unique()
+            cohort_size = len(cohort)
+
+            # Loop on actions
+            actions = (
+                window_statements.groupby(
+                    [
+                        "object.id",
+                        "object.definition.name",
+                        "context.extensions.http://lrs.learninglocker.net/define/extensions/info.event_name",
+                    ]
+                )["actor.account.name"]
+                .nunique()
+                .reset_index()
+            )
+            actions.rename(
+                columns={
+                    "object.id": "iri",
+                    "object.definition.name": "name",
+                    (
+                        "context.extensions.http://lrs.learninglocker.net"
+                        "/define/extensions/info.event_name"
+                    ): "module_type",
+                    "actor.account.name": "cohort",
+                },
+                inplace=True,
+            )
+
+            # Find active actions on the current window
+            for _, action in actions.iterrows():
+                # Check for action not recorded as an active action
+                if action.iri not in active_actions["iri"].values:
+                    # Record new active action
+                    if (
+                        0.1 * cohort_size <= action.cohort
+                        and action.cohort >= self.dynamic_cohort_min
+                    ):
+                        active_actions.loc[len(active_actions)] = action[
+                            ["iri", "name", "module_type"]
+                        ]
+
+            if len(active_actions) < self.active_actions_min:
+                since -= timedelta(days=1)  # step back from one day
+            else:
+                return SlidingWindowIndicator(
+                    window=Window(since=since, until=self.until),
+                    active_actions=self._compute_activation(
+                        statements, active_actions, cohort_size
+                    ),
+                    dynamic_cohort=cohort.tolist(),
+                )
+
+        return sliding_window
+
+    def dataframe_to_pydantic(self, model_class, dataframe):
+        """Convert a dataframe to a list of Pydantic model instances."""
+        columns = model_class.__annotations__.keys()
+
+        return [model_class(**row[columns]) for _, row in dataframe.iterrows()]
```

Comment:
Looks like it could be an utility function and not get tight to this indicator. WDYT?